OCPBUGS-9404: azure: skip LB creation when not needed

r4f4 opened the pull request:
When one creates an IPI Azure cluster with an `internal` publishing method, it creates a standard load balancer with an empty definition. This load balancer doesn't serve a purpose since the configuration is completely empty. Because it doesn't have a public IP address and backend pools it's not providing any outbound connectivity, and there are no frontend IP configurations for ingress connectivity to the cluster.

r4f4 commented:
/jira refresh

r4f4 commented:
/hold
Needs https://github.com/openshift/installer/pull/7064 for the CI jobs to run.

zaneb commented:
/retest
/hold cancel

patrickdillon commented on data/data/azure/vnet/public-lb.tf:
```diff
@@ -147,7 +148,7 @@ resource "azurerm_lb_outbound_rule" "public_lb_outbound_rule_v4" {
 
   name                    = "outbound-rule-v4"
-  loadbalancer_id         = azurerm_lb.public.id
+  loadbalancer_id         = azurerm_lb.public[0].id
```
Couldn't this resource be created when the load balancer doesn't exist (that is `var.azure_private == true && var.azure_outbound_user_defined_routing == true`) in which case I think we'd have an error?

r4f4 replied:
Fixed.

patrickdillon reviewed (commented):
> When one creates an IPI Azure cluster with an internal publishing method, it creates a standard load balancer with an empty definition. This load balancer doesn't serve a purpose since the configuration is completely empty. Because it doesn't have a public IP address and backend pools it's not providing any outbound connectivity, and there are no frontend IP configurations for ingress connectivity to the cluster.

It looks like the way we provide outbound connectivity to VMs when the install is private and not using user-defined routing, is by attaching an outbound rule to the load balancer:

https://github.com/openshift/installer/blob/ebabc7455f45b9f666aa5ea6cdd20c040a2cfa6e/data/data/azure/vnet/public-lb.tf#L146-L157

Which leaves me with the question: how are VMs getting outbound connectivity in this scenario? 

I was under the impression that we could only remove the LB if using user-defined routing.

r4f4 commented:
> We don't have any CI testing that performs a private install. I am floating [openshift/release#38408](https://github.com/openshift/release/pull/38408) as an option. Have you done any local testing of this? I'm not sure what the expectations are for the cloud provider.

It was pre-merge tested by @jinyunma before I addressed your comments. I haven't found the time to retest it yet after the update.

patrickdillon commented:
> > We don't have any CI testing that performs a private install. I am floating [openshift/release#38408](https://github.com/openshift/release/pull/38408) as an option. Have you done any local testing of this? I'm not sure what the expectations are for the cloud provider.
> 
> It was pre-merge tested by @jinyunma before I addressed your comments. I haven't found the time to retest it yet after the update.

Excellent! And sorry I missed that pre-merge testing in the bug. I'm not concerned about testing the recent update. Thank you!

patrickdillon commented:
@jinyunma Thanks for doing [the pre-merge testing](https://issues.redhat.com/browse/OCPBUGS-9404?focusedId=22074947&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-22074947) on this. I see that the testing was done with user-defined routing. That makes sense and it seems like we should be able to remove the LB in that case.

Are you able to test without user-defined routing? I believe in that case we still need to keep the LB, because the installer applies it's own rule to enable outbound internet access. If the install succeeds with the current changes, we want to double check to make sure Azure is not enabling [default outbound access](https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/default-outbound-access).

jinyunma commented:
> That makes sense and it seems like we should be able to remove the LB in that case.

yes, external LB is removed on fully private cluster (fully private cluster ( publish: Internal + outboundType: UserDefinedRouting), and no function impact.
 
> Are you able to test without user-defined routing? I believe in that case we still need to keep the LB, because the installer applies it's own rule to enable outbound internet access. If the install succeeds with the current changes, we want to double check to make sure Azure is not enabling [default outbound access](https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/default-outbound-access).

pre-merge testing on the latest update, install private cluster successfully ( publish: Internal  wihtout user-defined routing), external LB is created, and outbound rules is explicitly defined, using the frontend IP address of external LB for outbound.

```
# az network lb list -g jima7673-18b-slv6v-rg -otable
Location    Name                         ProvisioningState    ResourceGroup          ResourceGuid
----------  ---------------------------  -------------------  ---------------------  ------------------------------------
eastus      jima7673-18b-slv6v           Succeeded            jima7673-18b-slv6v-rg  8d5b10d7-9b15-4961-8968-00673ba6de33
eastus      jima7673-18b-slv6v-internal  Succeeded            jima7673-18b-slv6v-rg  a547ee2b-701c-4ada-9048-79df4b31ceac

# az network lb show --name jima7673-18b-slv6v -g jima7673-18b-slv6v-rg --query outboundRules
[
  {
    "allocatedOutboundPorts": 1024,
    "backendAddressPool": {
      "id": "/subscriptions/53b8f551-f0fc-4bea-8cba-6d1fefd54c8a/resourceGroups/jima7673-18b-slv6v-rg/providers/Microsoft.Network/loadBalancers/jima7673-18b-slv6v/backendAddressPools/jima7673-18b-slv6v",
      "resourceGroup": "jima7673-18b-slv6v-rg"
    },
    "enableTcpReset": false,
    "etag": "W/\"5b6a9854-dfa1-4972-999f-3475b70623e2\"",
    "frontendIpConfigurations": [
      {
        "id": "/subscriptions/53b8f551-f0fc-4bea-8cba-6d1fefd54c8a/resourceGroups/jima7673-18b-slv6v-rg/providers/Microsoft.Network/loadBalancers/jima7673-18b-slv6v/frontendIPConfigurations/public-lb-ip-v4",
        "resourceGroup": "jima7673-18b-slv6v-rg"
      }
    ],
    "id": "/subscriptions/53b8f551-f0fc-4bea-8cba-6d1fefd54c8a/resourceGroups/jima7673-18b-slv6v-rg/providers/Microsoft.Network/loadBalancers/jima7673-18b-slv6v/outboundRules/outbound-rule-v4",
    "idleTimeoutInMinutes": 4,
    "name": "outbound-rule-v4",
    "protocol": "All",
    "provisioningState": "Succeeded",
    "resourceGroup": "jima7673-18b-slv6v-rg",
    "type": "Microsoft.Network/loadBalancers/outboundRules"
  }
]
```

r4f4 commented:
@patrickdillon any other concerns?

patrickdillon commented on data/data/azure/vnet/public-lb.tf:
```diff
@@ -53,6 +53,7 @@ data "azurerm_public_ip" "cluster_public_ip_v6" {
 }
 
 resource "azurerm_lb" "public" {
+  count               = ! var.azure_private || ! var.azure_outbound_user_defined_routing ? 1 : 0
```
+1

r4f4 replied:
Done

patrickdillon commented:
/approve

r4f4 commented:
Update: no functional changes, just reusing a tf variable.

jhixson74 commented:
/lgtm

r4f4 commented:
/hold cancel

r4f4 commented:
/hold cancel

r4f4 commented:
/cherry-pick release-4.13
